Add cats module

ybasket opened the pull request:
This PR adds support for some basic type classes from `cats-core`:
- `Eq` for typesafe comparison
- `Show` for pretty printing
- `Hash` for abstracting of hashing of values

Corresponding enumeratum-style mixins are provided for these, supporting both `Enum` and `ValueEnum`. The second also comes in an extended version which provides an instance of `Order`.

Major use cases are generic derivation and operations on collections of (nested) enum values, for example sorting a list of errors by their (enum) priority. See `Readme.md` for small examples.

lloydmeta commented:
Nice! 

Do you think we should include a ScalaJS version of this too?

ybasket commented:
Appearently, the Slick tests are a bit flaky on Travis (it only failed on Scala 2.11 and I don't touch Slick at all) - ran the full test suite locally, all green. That seems also to be the reason the coverage report looks so bad.

lloydmeta commented on enumeratum-cats/src/main/scala/enumeratum/CatsEnum.scala:
```diff
@@ -0,0 +1,14 @@
+
+  implicit val eqInstance: Eq[A] = Eq.fromUniversalEquals[A]
+
+  implicit val showInstance: Show[A] = Show.fromToString[A]
```
Is this what we want? (e.g. rather than `_.entryName`)

lloydmeta commented on enumeratum-cats/src/main/scala/enumeratum/values/CatsValueEnum.scala:
```diff
@@ -0,0 +1,28 @@
+trait CatsValueEnum[ValueType, EntryType <: ValueEnumEntry[ValueType]] {
+  this: ValueEnum[ValueType, EntryType] =>
+
+  implicit val eqInstance: Eq[EntryType] = Eq.fromUniversalEquals[EntryType]
```
Would it make sense to have these depend on `Eq` on the `ValueType` and build an `Eq` for the `EntryType` using that?

ybasket replied:
Interesting question - I tend to say no. The idea of an enum is having distinct values and these are guaranteed (correct me if I'm wrong) by the way the macros work, requiring literals in the definition. So the only valid `Eq` instance would be the one differentiating all enum values and we achieve that already by `fromUniversalEquals`. One can still have a custom instance by overriding this one, it's a one-liner.

So I'd leave it as-is for now, but arguments welcome :)

ybasket commented:
@lloydmeta Maybe we can add a new section in `Readme.md` on how to implement a new module stating such best practices? I should have seen this on Circe, but I guess I'm not the only failing there ;)

Anyway, thanks for the helpful and quick feedback! I updated the PR to provide this and also added some Scaladoc and a Readme section on how to use the Cats module without inheritance. These objects also provide methods to create instances of (for example) `Eq` based on a custom `Eq`, so now one can now more easily decide to use the default or something different.

lloydmeta commented:
Released as 1.5.14; should be up on Maven Central soon :) Thanks again !
